# Introduction to Dataset
## Data Description
The dataset we are using is comprised of 2.5 years of insulin dosage, blood glucose (bg), and Estimated Variability of Glucose (EVG). This data was collected from a type 1 diabetic's insulin pump and continuous glucose monitor (CGM). Unfortunately this data isnt entirely continous over the 2.5 year span and has gaps in some instances. This data has been exported directly from 30 day span .csv files containing 3 tables one for insulin dosage (bolus), bg, and EVG data. We split these files into the 3 respective tables and saved them to new tables containing the full time-span of data, these files are titles Bolus.csv, BG.csv, and EVG.csv. Here I will detail the contents of these files.

1. Bolus
    * Features:
        * Type: Type of bolus event (Always 'Bolus').
        * BolusType: Describes the wayin which the Bolus was used (categorical).
        * BolusDeliveryMethod: Method used to deliver the bolus (Auto or Standard)(categorical).
        * BG (mg/dL): Blood glucose levels at the time of bolus administration (continuous data).
        * SerialNumber: Device identifier.
        * CompletionDateTime: Timestamp when the dosing was completed.
        * InsulinDelivered: The standard unit measure for the amount of insulin delivered (continuous data).
        * FoodDelivered: Insulin delivered for food consumption (continuous data).
        * CorrectionDelivered: Insulin delivered for BG correction (continuous data).
        * CompletionStatusDesc: Status description of the dosage (categorical).
        * BolexStartDateTime: Not Used In Export.
        * BolexCompletionDateTime: Not Used In Export.
        * BolexInsulinDelivered: Not Used In Export.
        * BolexCompletionStatusDesc: Not Used In Export.
        * StandardPercent: (Always 100).
        * Duration (mins):(Always 0).
        * CarbSize: The amount of carbohydrates consumed in grams (continuous data).
        * TargetBG (mg/dL): Target blood glucose level for the subject in milligrams per deciliter (continuous data).
        * CorrectionFactor: Insulin sensitivity factor (continuous data).
        * CarbRatio: Insulin-to-carbohydrate ratio (continuous data).
    * Notes:
        * This table holds the most detailed records for insulin dosing decisions, food intake, and blood glucose corrections.
        * This table will be very useful in modeling relationships between carbohydrate intake, insulin dosage, and BG levels.
        * Bolus (Def: A large single doseage of insuline to lower a bloodsugar rise) refers to the device/method of insuline delivery. Which is automated through a pocket size device with a refillable tank of insulin.
2. EVG Table
    * Features:
        * DeviceType: Type/Name of the device used to record the event.
        * SerialNumber: Device identifier.
        * Description: A text description of the type of data recorded (Always EVG).
        * EventDateTime: Timestamp of when the measurement was recorded.
        * Readings (mg/dL): The estimated glucose level at the recorded time, in milligrams per deciliter (mg/dL).
    * Notes: 
        * The EVG data is collected directly from a CGM.
        * EVG is designed for tracking overall trends and patterns in glucose levels rather than moment-to-moment decisions.
        * Useful for identifying time-in-range, glucose variability, and predicting hypo/hyperglycemia over time.
        
3. BG Table
    * Features:
        * DeviceType: Type/Name of device used to measure blood glucose.
        * SerialNumber: Device identifier.
        * Description: A text description of the type of data recorded (Always BG)
        * EventDateTime: Timestamp of the blood glucose measurement.
        * BG (mg/dL): Blood glucose levels in milligrams per deciliter (continuous data).
        * Note: Additional notes field (Always Blank).
    * Notes:
        * This data is manually entered and is recorded through a glucometer (Finger Prick).
        * The BG measurement reflects the actual glucose levels at the time of measurement.


### General Dataset Properties

* Total \# of Features:
    * Bolus: 18 features
    * EVG: 5 features
    * BG: 6 features
* \# of Usable/Useful Features: (!!!!! Update possibly)
    * Bolus: 12 features
    * EVG: 2 features
    * BG: 2 features
* Table Sizes:
    * Bolus: 11,348 Records
    * EVG: 191,781 Records
    * BG: 2,832 Records
* Unique Dates: 
    * Bolus: 599 Days
    * EVG: 666 Days
    * BG: 663 Days
* Intersecting Dates: 599



## Preprocessing

### Methodology for Data Preprocessing
Given the nature of our timeseries problem after initial testing we identified some issues with the bolus data not being recorded at consistent intervals and only when administering insulin. To remedy this given that our EVG data is continous we decided to fill the gaps in our data with blank bolus records, only adding a 'Type' field equal to 'EVG', to denote the different types of records, and also adding the BG readings while setting the rest of the numeric fields to 0 and categorical fields to ''. Below I detail the preprocessing process.

**1. Data Import and Initial Setup**
* Loading Data:
  * We decided to use the `EVG.csv` and `Bolus.csv` files as pandas DataFrames.
  * In the function `convert_data(df):` we prepare the `EVG` data by mapping it to match the `Bolus` table structure for consistent formatting across datasets.
* Merging Tables:
  * We merged the `EVG` and `Bolus` tables into a single dataset using pandas' `concat` function filling the gaps in time.

**2. Data Cleaning**
* Datetime Conversion:
  * Converted the `CompletionDateTime` column to the dataframe index.
  * Rounded the `CompletionDateTime` to the nearest 5-minute interval to standardize timestamps.
  * Identified and removed rows where the `CompletionDateTime` overlapped with `CGM` data to avoid redundancy or conflicting records.
  * Dropped all data before 2024 to ensure continous data throughout the set.

**3. Data Transformation**
  * Applied one-hot encoding to categorical columns, such as `Type`, `BolusType`, `BolusDeliveryMethod`, and `CompletionStatusDesc`, to prepare them for machine learning models.
    * We considered using encodings to map categorical values but resorted to one-hot encoding as it was more familiar to us.
  * Removed columns that were not relevant or consistently populated (`SerialNumber`, `BolexStartDateTime`, `BolexCompletionDateTime`, etc.).
  * Converted boolean columns to integer type for min-max scaling.
  * Re-ordered columns and set the Date value as the Dataframe index.
  * Ensured no missing values (`NaN`) remained in the dataset.
  * Normalized the data using `MinMaxScaler` to normalize numeric fields to a range of [0, 1].


### **Function Descriptions**

**1. `round_to_nearest_5_minutes(df)`**
* Purpose: To ensure all timestamps (`CompletionDateTime`) in the dataset are standardized by rounding them to the nearest 5-minute interval.
* Steps:
  1. Converts the `CompletionDateTime` column to datetime format.
  2. Drops rows with invalid or missing datetime values.
  3. Rounds valid `CompletionDateTime` values to the nearest 5-minute interval.


**2. `remove_cgm_on_time_overlap(df)`**
* Purpose: To remove redundant or conflicting data points from the dataset when multiple records share the same timestamp.
* Steps:
  1. Identifies duplicate `CompletionDateTime` entries across the dataset.
  2. Filters out rows where the `Type` is `CGM` and the timestamp overlaps with other records.


**3. `convert_data(df)`**
* Purpose: To map the structure of the `EVG` table to match the `Bolus` table schema, enabling consistent formatting and integration.
* Steps:
  1. Creates a template `Bolus` table structure with predefined columns.
  2. Renames relevant columns in the `EVG` data to align with the `Bolus` schema.
  3. Adds missing columns to the mapped `EVG` table, using default values for both numeric and categorical fields.
  4. Reorders the columns to match the `Bolus` table structure.


**4. `onehot_encoding(df, columns)`**
* Purpose: To transform specified categorical columns into one-hot encoded vectors.
* Steps:
  1. Uses pandas' `get_dummies()` method to create binary indicator columns for each category in the specified columns.
  2. Appends these new columns to the DataFrame.



In [ ]:
import pandas as pd
from sklearn.preprocessing import MinMaxScaler, StandardScaler, RobustScaler



def round_to_nearest_5_minutes(df):
    # Ensure CompletionDateTime is in datetime format
    df['CompletionDateTime'] = pd.to_datetime(df['CompletionDateTime'], errors='coerce')
    # Drop rows with invalid datetime entries
    df = df.dropna(subset=['CompletionDateTime'])
    # Round CompletionDateTime to the nearest 5 minutes
    df['CompletionDateTime'] = df['CompletionDateTime'].dt.round('5min')
    return df

def remove_cgm_on_time_overlap(df):
    # Identify duplicates based on CompletionDateTime
    duplicate_times = df[df.duplicated('CompletionDateTime', keep=False)]
    # Filter out rows where Type is EVG and CompletionDateTime is duplicated
    df = df[~((df['CompletionDateTime'].isin(duplicate_times['CompletionDateTime'])) & (df['Type'] == 'CGM'))]
    
    return df

def convert_data(df):
    # Example Bolus table columns
    bolus_columns = [
        'Type', 'BolusType', 'BolusDeliveryMethod', 'BG (mg/dL)', 'SerialNumber',
        'CompletionDateTime', 'InsulinDelivered', 'FoodDelivered', 'CorrectionDelivered',
        'CompletionStatusDesc', 'BolexStartDateTime', 'BolexCompletionDateTime', 
        'BolexInsulinDelivered', 'BolexCompletionStatusDesc', 'StandardPercent', 
        'Duration (mins)', 'CarbSize', 'TargetBG (mg/dL)', 'CorrectionFactor', 'CarbRatio'
    ]

    # 1. Create a template DataFrame for Bolus table with default values
    template_bolus = pd.DataFrame(columns=bolus_columns)

    # Default values for missing fields
    default_numeric = 0
    default_categorical = ''

    # 2. Map EVG fields to Bolus fields
    mapped_bolus = evg_df.rename(columns={
        'Readings (mg/dL)': 'BG (mg/dL)',
        'SerialNumber': 'SerialNumber',
        'EventDateTime': 'CompletionDateTime'
    })

    # 3. Add missing columns with default values
    for col in bolus_columns:
        if col not in mapped_bolus.columns:
            if col in ['BG (mg/dL)', 'InsulinDelivered', 'FoodDelivered', 'CorrectionDelivered', 
                    'StandardPercent', 'Duration (mins)', 'CarbSize', 'TargetBG (mg/dL)', 
                    'CorrectionFactor', 'CarbRatio']:  # Numeric fields
                mapped_bolus[col] = default_numeric
            elif col == 'Type':
                mapped_bolus[col] = 'CGM'
            else:  # Categorical fields
                mapped_bolus[col] = default_categorical

    # 4. Reorder columns to match the Bolus table structure
    mapped_bolus = mapped_bolus[bolus_columns]

    # Resulting Bolus DataFrame
    #print(mapped_bolus)
    return mapped_bolus



def onehot_encoding(df, columns):
    df = pd.get_dummies(df, columns=columns)
    return df




# Load the CSV files into pandas DataFrames
#file_path1 = './DataTables/BG.csv'  # Replace with the actual path for Table 1
file_path2 = './DataTables/EVG.csv'  # Replace with the actual path for Table 2
file_path3 = './DataTables/Bolus.csv'  # Replace with the actual path for Table 3

# Load the data
#df1 = pd.read_csv(file_path1)
evg_df = pd.read_csv(file_path2)
bolus_df = pd.read_csv(file_path3)

evg_mapped = convert_data(evg_df)

df = pd.concat([bolus_df, evg_mapped], ignore_index=True)
#print(evg_mapped)

# Round CompletionDateTime to the nearest 5 minutes
df = round_to_nearest_5_minutes(df)

# Sort the DataFrame by CompletionDateTime
df = df.sort_values(by='CompletionDateTime').reset_index(drop=True)

df = remove_cgm_on_time_overlap(df)
columns=['Type', 'BolusType', 'BolusDeliveryMethod', 'CompletionStatusDesc']
# Apply the function
df = onehot_encoding(df, columns)
columns=['SerialNumber', 'BolexStartDateTime', 'BolexCompletionDateTime', 'BolexInsulinDelivered', 'BolexCompletionStatusDesc']
df = df.drop(columns=columns)
bool_columns = df.select_dtypes(include='bool').columns
df[bool_columns] = df[bool_columns].astype(int)

df.rename(columns={'CompletionDateTime': 'Date'}, inplace=True)

# Move the 'Date' column to the first position
first_col = df.pop('Date')  # Remove 'Date' column
df.insert(0, 'Date', first_col)  # Reinsert it at position 0

# Display the DataFrame

# Check the resulting DataFrame
df.to_csv('test.csv', index=False)





# Check for NaN values; there are none
print("NaN values:", df.isnull().values.any())
# Organize data, make date into an index
df.set_index("Date", inplace=True)
# Note: we are not using any categorical data
values = df.values
# specify columns to plot
#groups = [0, 1, 2, 3]
#i = 1
# plot each column
#pyplot.figure()
#for group in groups:
#	pyplot.subplot(len(groups), 1, i)
#	pyplot.plot(values[:, group])
#	pyplot.title(df.columns[group], y=0.5, loc='right')
#	i += 1
#pyplot.show()
# normalize features
scaler = MinMaxScaler(feature_range=(0, 1))
scaled = scaler.fit_transform(values)
print(scaled)

# Choosing a Model

### Baseline Examples

### Introduction to Time Series

### From Neural Networks to Recurrent Neural Networks

### Recurrent Neural Networks Improved: Long-Short Term Memory Model

# Tutorial by Jason Brownlee

Here we simply use the Time Series to Supervised Set formatter written by Jason:

In [132]:
from pandas import DataFrame
from pandas import concat
 
def series_to_supervised(data, n_in=1, n_out=1, dropnan=True):
	"""
	Frame a time series as a supervised learning dataset.
	Arguments:
		data: Sequence of observations as a list or NumPy array.
		n_in: Number of lag observations as input (X).
		n_out: Number of observations as output (y).
		dropnan: Boolean whether or not to drop rows with NaN values.
	Returns:
		Pandas DataFrame of series framed for supervised learning.
	"""
	n_vars = 1 if type(data) is list else data.shape[1]
	df = DataFrame(data)
	cols, names = list(), list()
	# input sequence (t-n, ... t-1)
	for i in range(n_in, 0, -1):
		cols.append(df.shift(i))
		names += [('var%d(t-%d)' % (j+1, i)) for j in range(n_vars)]
	# forecast sequence (t, t+1, ... t+n)
	for i in range(0, n_out):
		cols.append(df.shift(-i))
		if i == 0:
			names += [('var%d(t)' % (j+1)) for j in range(n_vars)]
		else:
			names += [('var%d(t+%d)' % (j+1, i)) for j in range(n_vars)]
	# put it all together
	agg = concat(cols, axis=1)
	agg.columns = names
	# drop rows with NaN values
	if dropnan:
		agg.dropna(inplace=True)
	return agg

# Implementation of Tutorial

### Selecting Variables From Our Dataset

### Final Data Preprocessing Step

Loading data from our previously-made Bolus.csv table into a Pandas dataframe:

In [133]:
import pandas as pd
from sklearn.preprocessing import MinMaxScaler

# Load Bolus Data
file_path = './DataTables/Bolus.csv'
df = pd.read_csv(file_path)
df['CompletionDateTime'] = pd.to_datetime(df['CompletionDateTime'])
df = df.sort_values(by='CompletionDateTime')

Here we prepare and clean our data:

In [134]:
# Cull Featuers We're Not Using
df = df[["BG (mg/dL)", "CompletionDateTime", "InsulinDelivered", "FoodDelivered", "CarbSize"]]

# Check for NaN values
print("NaN values:", df.isnull().values.any())

# Organize data, make dataframe indexable by date
df.columns = ["BG", "Date", "InsulinDelivered", "FoodDelivered", "CarbSize"]
columns_titles = ["Date", "BG", "InsulinDelivered", "FoodDelivered", "CarbSize"]
df = df.reindex(columns=columns_titles)
df.set_index("Date", inplace=True)
print(df.head())

# normalize features
values = df.values
scaler = MinMaxScaler(feature_range=(0, 1))
scaled = scaler.fit_transform(values)

NaN values: False
                      BG  InsulinDelivered  FoodDelivered  CarbSize
Date                                                               
2022-04-27 11:38:14    0              5.00           5.00        75
2022-04-27 20:34:41  132              1.96           1.67        25
2022-04-28 06:43:30    0              1.07           1.07        16
2022-04-28 11:39:14    0              5.00           5.00        75
2022-04-28 18:09:54    0              1.07           1.07        16


### Multiple Timesteps

The "n_in" parameter in Jason's series_to_supervised() function allows us to transform the data such that each feature's time series is matched by *n* other time series of the feature which are offset by (t-1, t-2, ..., t-n) timesteps. If we leave it at default, we can then see, once we print the head of the dataframe, that it has output two timesteps for each variable. For variable 1, our blood glucose level as shown in the code cell above, there is var1(t-1) and var1(t). If you run the code cell below, you will notice that the time series var1(t-1) is the same series as var1(t) except that the entire series is offset one step into the past. This is the same for every one of the variables / features.

It will be more convenient for us to do this later, but if we culled the last three timesteps, we would be left with a dataframe containing four of previous timestep series for all of our features (var1(t-1), var2(t-2), var3(t-3), var4(t-4)) — our input variables — and one timestep series for the current blood glucose level (var1(t)) — our output variable.

In [135]:
# Transform to supervised set with one timestep
df_reframed = series_to_supervised(scaled, n_in=1, n_out=1)
print(df_reframed.head())

   var1(t-1)  var2(t-1)  var3(t-1)  var4(t-1)  var1(t)   var2(t)   var3(t)  \
1       0.00   0.267380   0.267380   0.166667     0.22  0.104813  0.089305   
2       0.22   0.104813   0.089305   0.055556     0.00  0.057219  0.057219   
3       0.00   0.057219   0.057219   0.035556     0.00  0.267380  0.267380   
4       0.00   0.267380   0.267380   0.166667     0.00  0.057219  0.057219   
5       0.00   0.057219   0.057219   0.035556     0.00  0.142781  0.142781   

    var4(t)  
1  0.055556  
2  0.035556  
3  0.166667  
4  0.035556  
5  0.088889  


If we set "n_in" parameter for more timesteps, the result is the same as above but with more timesteps. For our initial attempt, we will use 10 timesteps because we would later like to be able to use data from the last several hours as input into the LSTM to get our predicted blood-glucose level. "n_out" is set to 1 because we want to predict our current blood glucose level.

We *would* like to be able to predict future blood glucose levels (which would be n_out >= 2), but because our data was not recorded at regular intervals, timestep predictions into the future would tell us "how much?" but not exactly "when?". If our data were recorded in regular one hour intervals, we could then rest assured that future timestep predictions would be *n* number of hours into the future precisely. This is a limitation in our dataset or, at least, in the tutorial we have followed to implement our LSTM. The tutorial was working with data on regular intervals. We are not.

In [136]:
# Transform to supervised set with multiple timesteps
n_timesteps = 10
df_reframed = series_to_supervised(scaled, n_in=n_timesteps, n_out=1)
print(df_reframed.head())

    var1(t-10)  var2(t-10)  var3(t-10)  var4(t-10)  var1(t-9)  var2(t-9)  \
10        0.00    0.267380    0.267380    0.166667       0.22   0.104813   
11        0.22    0.104813    0.089305    0.055556       0.00   0.057219   
12        0.00    0.057219    0.057219    0.035556       0.00   0.267380   
13        0.00    0.267380    0.267380    0.166667       0.00   0.057219   
14        0.00    0.057219    0.057219    0.035556       0.00   0.142781   

    var3(t-9)  var4(t-9)  var1(t-8)  var2(t-8)  ...  var3(t-2)  var4(t-2)  \
10   0.089305   0.055556        0.0   0.057219  ...   0.246524   0.166667   
11   0.057219   0.035556        0.0   0.267380  ...   0.000000   0.000000   
12   0.267380   0.166667        0.0   0.057219  ...   0.106952   0.066667   
13   0.057219   0.035556        0.0   0.142781  ...   0.000000   0.000000   
14   0.142781   0.088889        0.0   0.057219  ...   0.000000   0.000000   

    var1(t-1)  var2(t-1)  var3(t-1)  var4(t-1)   var1(t)   var2(t)   var3(t)  \


### Split Test and Train Sets

Here we take data only from Fed 2024 through Oct 2024 because that time range in the Bolus dataset is the largest period of time for which there are not massive holes in the record. We split our data into a training set, comprised of the first half of our time range, and a testing set, comprised of the second half of our time range. We choose to split it this way rather than use standard shuffling methods of generating the sets because our data is only useful insofar as it records a time series. That is, if we scrambled the time series, the data would be relatively meaningless, yielding no info about the relative order nor trends in the data over time.

In [137]:
# Split into train and test sets
values = df_reframed.values
index_2024 = 6893
index_last = 11349
index_midpoint = index_2024 + ((index_last - index_2024) // 2)
train = values[index_2024:index_midpoint,:]
test = values[index_midpoint:,:]

Here we cull all the current timestep features except variable 1, which is the blood glucose level variable we are trying to predict.

In [138]:
# Split into input and outputs
n_features = 4
n_obs = n_timesteps * n_features
train_X, train_y = train[:, 0:n_obs], train[:, -n_features]
test_X, test_y = test[:, 0:n_obs], test[:, -n_features]

# reshape input to be 3D [samples, timesteps, features]
train_X = train_X.reshape((train_X.shape[0], n_timesteps, n_features))
test_X = test_X.reshape((test_X.shape[0], n_timesteps, n_features))

### Create and Fit Model

In [139]:
from keras.models import Sequential
from keras.layers import Input, Dense
from keras.layers import LSTM

# design LSTM
print(train_X.shape)
model = Sequential()
model.add(Input(shape=(train_X.shape[1], train_X.shape[2])))
model.add(LSTM(50))
model.add(Dense(1))
model.compile(loss='mae', optimizer='adam')

# fit LSTM
history = model.fit(train_X, train_y, epochs=50, batch_size=72, validation_data=(test_X, test_y), verbose=2, shuffle=False)

(2228, 10, 4)
Epoch 1/50


31/31 - 1s - 36ms/step - loss: 0.1840 - val_loss: 0.1288
Epoch 2/50
31/31 - 0s - 4ms/step - loss: 0.1288 - val_loss: 0.1245
Epoch 3/50
31/31 - 0s - 4ms/step - loss: 0.1270 - val_loss: 0.1236
Epoch 4/50
31/31 - 0s - 4ms/step - loss: 0.1263 - val_loss: 0.1229
Epoch 5/50
31/31 - 0s - 4ms/step - loss: 0.1258 - val_loss: 0.1223
Epoch 6/50
31/31 - 0s - 4ms/step - loss: 0.1253 - val_loss: 0.1219
Epoch 7/50
31/31 - 0s - 4ms/step - loss: 0.1250 - val_loss: 0.1215
Epoch 8/50
31/31 - 0s - 4ms/step - loss: 0.1245 - val_loss: 0.1211
Epoch 9/50
31/31 - 0s - 4ms/step - loss: 0.1243 - val_loss: 0.1206
Epoch 10/50
31/31 - 0s - 4ms/step - loss: 0.1240 - val_loss: 0.1202
Epoch 11/50
31/31 - 0s - 4ms/step - loss: 0.1236 - val_loss: 0.1198
Epoch 12/50
31/31 - 0s - 4ms/step - loss: 0.1234 - val_loss: 0.1194
Epoch 13/50
31/31 - 0s - 4ms/step - loss: 0.1229 - val_loss: 0.1190
Epoch 14/50
31/31 - 0s - 4ms/step - loss: 0.1228 - val_loss: 0.1188
Epoch 15/50
31/31 - 0s - 4ms/step - loss: 0.1225 - val_loss: 0.1185

### Evaluate Model

In his original tutorial, Jason used RMSE as his measurement of the performance of the dataset. MAE is, however, significantly easier to understand intuitively. So, we wanted to use MAE, but RMSE is better at finding significant errors, which is especially important since we are trying to predict human blood glucose levels. Luckily, we can use both to gain a better understanding of our dataset.

According to EUMeTrain.org, an international training project for the development of training material and training support in the field of satellite meteorology:

"The MAE and the RMSE can be used together to diagnose the variation in the errors in a set of forecasts. The RMSE will always be larger or equal to the MAE; the greater difference between them, the greater the variance in the individual errors in the sample. If the RMSE=MAE, then all the errors are of the same magnitude. Both the MAE and RMSE can range from 0 to ∞. They are negatively-oriented scores: Lower values are better."

So in the code cell below, we calculated MAE, RMSE, and the resulting Variance score.

In [140]:
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import root_mean_squared_error
import numpy as np

# make a prediction
yhat = model.predict(test_X)
test_X = test_X.reshape((test_X.shape[0], n_timesteps*n_features))
# invert scaling for forecast
inv_yhat = np.concatenate((yhat, test_X[:, -(n_features-1):]), axis=1)
inv_yhat = scaler.inverse_transform(inv_yhat)
inv_yhat = inv_yhat[:,0]
# invert scaling for actual
test_y = test_y.reshape((len(test_y), 1))
inv_y = np.concatenate((test_y, test_X[:, -(n_features-1):]), axis=1)
inv_y = scaler.inverse_transform(inv_y)
inv_y = inv_y[:,0]
# calculate MAE
mae = np.sqrt(mean_absolute_error(inv_y, inv_yhat))
print('Test MAE: %.3f' % mae)
rmse = np.sqrt(root_mean_squared_error(inv_y, inv_yhat))
print('Test RMSE: %.3f' % rmse)
variance = rmse - mae
print('Test Variance: %.3f' % variance)

70/70 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step
Test MAE: 8.372
Test RMSE: 10.045
Test Variance: 1.673


Running the code cell above results in an MAE of 8.372, an RMSE of 10.045, and a variance of 1.673. This variance is very low, and so it is unlikely that very large errors have occured. MAE tells us that the average difference between the forecasted blood glucose level and the observed was 8.3 mg/dL. The monitor our data came from records BG in a range from 0 to 400 mg/dL. 8.372 is only about 2.1% ((8.372 / 400) * 100 = 2.093). So, on average, our model's predicted BG level is only about 2.1% off of the actual level.

# Tuning

### Are we overfitting? Underfitting? How do we tell?

### Which input variables from out dataset are the best predictors?

### What is the optimal number of timestamps?

### What are the best LSTM hyperparameters?

In [ ]:
# From tutorial: "Remember that the internal state of the LSTM in Keras is reset at the end of each batch, so an internal state that is a function of a number of days may be helpful (try testing this)"

# Gird Search implementation like the one explained in Lecture (takes 15 mins to run currently)
def GridSearchHyperparameters(num_neurons, num_epochs, batch_size, validation_data):
    neurons_best = None
    epochs_best = None
    batch_size_best = None
    mae_best = 1000000000000
    for neurons in num_neurons:
        for epochs in num_epochs:
            for batches in batch_size:
                print('------------------------------------')
                print("neurons: " + str(neurons), "epochs: " + str(epochs), "batches: " + str(batches))
                # reset data
                test_X = validation_data[0]
                test_y = validation_data[1]
                # design LSTM
                model = Sequential()
                model.add(Input(shape=(train_X.shape[1], train_X.shape[2])))
                model.add(LSTM(neurons))
                model.add(Dense(1))
                model.compile(loss='mae', optimizer='adam')
                # fit LSTM
                model.fit(train_X, train_y, epochs=epochs, batch_size=batches, validation_data=validation_data, verbose=0, shuffle=False)
                # make a prediction
                yhat = model.predict(test_X)
                test_X = test_X.reshape((test_X.shape[0], n_timesteps*n_features))
                # invert scaling for forecast
                inv_yhat = np.concatenate((yhat, test_X[:, -(n_features-1):]), axis=1)
                inv_yhat = scaler.inverse_transform(inv_yhat)
                inv_yhat = inv_yhat[:,0]
                # invert scaling for actual
                test_y = test_y.reshape((len(test_y), 1))
                inv_y = np.concatenate((test_y, test_X[:, -(n_features-1):]), axis=1)
                inv_y = scaler.inverse_transform(inv_y)
                inv_y = inv_y[:,0]
                # calculate MAE
                mae = np.sqrt(mean_absolute_error(inv_y, inv_yhat))
                # best MAE decision
                if mae < mae_best:
                    neurons_best = neurons
                    epochs_best = epochs
                    batch_size_best = batches
                    mae_best = mae
    return mae_best, neurons_best, epochs_best, batch_size_best


# Split into input and outputs
n_features = 4
n_obs = n_timesteps * n_features
train_X, train_y = train[:, 0:n_obs], train[:, -n_features]
test_X, test_y = test[:, 0:n_obs], test[:, -n_features]

# reshape input to be 3D [samples, timesteps, features]
train_X = train_X.reshape((train_X.shape[0], n_timesteps, n_features))
test_X = test_X.reshape((test_X.shape[0], n_timesteps, n_features))
print(GridSearchHyperparameters(num_neurons=[5,10,25,50],
                                num_epochs=[10, 20, 40, 80],
                                batch_size=[10, 20, 40, 80],
                                validation_data=(test_X, test_y))
      )

# Results

### Summary

### Limitations

### Credits
* Jason Brownlee's tutorial on how to create a multivariate LSTM in Python @ https://machinelearningmastery.com/multivariate-time-series-forecasting-lstms-keras/
* "The AI Hacker" on YouTube for his informative guide to RNNs @ https://youtu.be/LHXXI4-IEns?si=C4FoMP-Bv4xywsHZ
* EUMeTrain.org for their explanation of MAE, RMSE, and how to assess variance between them @ https://resources.eumetrain.org/data/4/451/english/msg/ver_cont_var/uos3/uos3_ko1.htm


### Use of AI